### PR TITLE
Warning threshold rule

### PR DESCRIPTION
**Pull request by woodhamgh**
# What <h3> 
Rule to check the number of warnings thrown by Swiftlint and to throw an error if the number of warnings is greater than the threshold.

# How <h3> 
New configuration parameter which adds and error into the list of violations if the threshold is broken. Done in the lint command 

Can be configured in the config file with
 `warning_threshold: 5`

# Who <h3>
https://github.com/woodhamgh

# Why <h3> 
This allows for some warnings to go through without causing build failures but limits the number of warnings within the project. Some workplaces might have their own reasons for this or for how many they are happy to allow through.

Issue where request was made:
https://github.com/realm/SwiftLint/issues/696

### CODE REVIEW

**Comment by masters3d** [comment]
```
    ✗ testSwiftLintLints, failed - Function body should span 40 lines or less excluding comments and whitespace: currently spans 41 lines

```

File: Source/SwiftLintFramework/Models/Configuration.swift [hunk 1]
```diff
@@ -72,6 +75,11 @@ public struct Configuration: Equatable {
             return nil
         }
 
+        // set the config threshold to the threshold provided in the config file
+        if let warningThreshold = warningThreshold {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/Configuration.swift]: this conditional is unnecessary

File: Source/SwiftLintFramework/Models/Configuration.swift [hunk 2]
```diff
@@ -141,6 +150,7 @@ public struct Configuration: Equatable {
             whitelistRules: defaultStringArray(dict[ConfigurationKey.WhitelistRules.rawValue]),
             included: defaultStringArray(dict[ConfigurationKey.Included.rawValue]),
             excluded: defaultStringArray(dict[ConfigurationKey.Excluded.rawValue]),
+            warningThreshold: dict[ConfigurationKey.WarningThreshold.rawValue] as? Int ?? nil,
```
Comment (jpsim) [inline Source/SwiftLintFramework/Models/Configuration.swift]: the `?? nil` is unnecessary.

File: Source/swiftlint/Commands/LintCommand.swift [hunk 3]
```diff
@@ -55,7 +55,10 @@ struct LintCommand: CommandType {
             violations += currentViolations
             reporter.reportViolations(currentViolations, realtimeCondition: true)
         }.flatMap { files in
-            reporter.reportViolations(violations, realtimeCondition: false)
+            if isWarningThresholdBroken(configuration, violations: violations) {
+                violations.append(createThresholdViolation(configuration.warningThreshold!))
+            }
+            reporter.reportViolations(violations, realtimeCondition: true)
```
Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: why did you change `realtimeCondition` from `false` to `true`?

File: Source/swiftlint/Commands/LintCommand.swift [hunk 4]
```diff
@@ -119,3 +122,27 @@ struct LintOptions: OptionsType {
             <*> mode <| quietOption(action: "linting")
     }
 }
+
+private func isWarningThresholdBroken(configuration: Configuration,
+                                      violations: [StyleViolation]) -> Bool {
+    guard let warningThreshold = configuration.warningThreshold else { return false }
+    let numberOfWarningViolations = violations.filter({ $0.severity == .Warning}).count
+    if numberOfWarningViolations >= warningThreshold {
```
Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: these 4 lines should just be `return numberOfWarningViolations >= warningThreshold`.

File: Source/swiftlint/Commands/LintCommand.swift [hunk 5]
```diff
@@ -119,3 +122,27 @@ struct LintOptions: OptionsType {
...
+        identifier: "warning_threshold",
+        name: "Warning Threshold",
+        description: "Number of warnings thrown is above the threshold."
+    )
+    let location = Location(file: "", line: 0, character: 0)
+    return StyleViolation(
+        ruleDescription: description,
+        severity: ViolationSeverity.Error,
+        location: location,
+        reason: "Number of warnings exceeded threshold of \(threshold)")
```
Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: Missing period at the end of the string.

File: Source/swiftlint/Commands/LintCommand.swift [hunk 6]
```diff
@@ -119,3 +122,27 @@ struct LintOptions: OptionsType {
...
+private func createThresholdViolation(threshold: Int) -> StyleViolation {
+    let description = RuleDescription(
+        identifier: "warning_threshold",
+        name: "Warning Threshold",
+        description: "Number of warnings thrown is above the threshold."
+    )
+    let location = Location(file: "", line: 0, character: 0)
+    return StyleViolation(
+        ruleDescription: description,
+        severity: ViolationSeverity.Error,
```
Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: `severity: .Error` please

**Comment by jpsim** [comment]
Could you please add a changelog entry? Also, a test that verifies that an error is reported if the warning threshold is exceeded would be nice.

File: Source/swiftlint/Commands/LintCommand.swift (hunk 3 above)
Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: Hey, when it was false the new error would not show up in the reporting. Im guessing because the report was already generated and so it didn't generate again without True.

Happy to change if you know how this could be done differently

File: Source/swiftlint/Commands/LintCommand.swift (hunk 4 above)
Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: done

File: Source/swiftlint/Commands/LintCommand.swift (hunk 5 above)
Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: done

File: Source/swiftlint/Commands/LintCommand.swift (hunk 6 above)
Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: done, sorry

File: Source/SwiftLintFramework/Models/Configuration.swift (hunk 1 above)
Comment (woodhamgh) [inline Source/SwiftLintFramework/Models/Configuration.swift]: done

File: Source/SwiftLintFramework/Models/Configuration.swift (hunk 2 above)
Comment (woodhamgh) [inline Source/SwiftLintFramework/Models/Configuration.swift]: done

File: Source/swiftlint/Commands/LintCommand.swift (hunk 3 above)
Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: Could you please double-check that? It should definitely be reported.

Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: Yup so changed to false and the error no longer shows up. It still adds to the array but then I get 

"Command /bin/sh failed with exit code 2" whereas if it is true I get an error shown and then the done linting comment (in xcode)

Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: Ah, I know what's happening... this code should be this:

```swift
if isWarningThresholdBroken(configuration, violations: violations) {
  violations.append(createThresholdViolation(configuration.warningThreshold!))
  if reporter.isRealtime {
    reporter.reportViolations([violations.last!], realtimeCondition: true)
  }
}
reporter.reportViolations(violations, realtimeCondition: false)
```

Though this is pretty confusing and could stand to be refactored, but not in this PR.

Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: What would you like me to do? Cause i think this func is now at its 40line limit

Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: Actually, the following works and keeps the function under 40 lines:

```swift
if isWarningThresholdBroken(configuration, violations: violations) {
  violations.append(createThresholdViolation(configuration.warningThreshold!))
  reporter.reportViolations([violations.last!], realtimeCondition: true)
}
reporter.reportViolations(violations, realtimeCondition: false)
```

Comment (woodhamgh) [inline Source/swiftlint/Commands/LintCommand.swift]: Done. With the tests I cant seem to find anywhere there are tests for the LintCommand.

Would you like me to create a test class in this swiftlint group or make one in the SwiftLintFrameworkTests

Comment (jpsim) [inline Source/swiftlint/Commands/LintCommand.swift]: Oh, I just realized that this is in `LintCommand`, which is in the `swiftlint` target, which isn't tested...

We only have unit tests set up for SwiftLintFramework.

Please disregard my comment about adding a test for this then.

**Comment by jpsim** [comment]
Passed CI, this is going in. Big thank you @woodhamgh!

**Comment by woodhamgh** [comment]
@jpsim Wooh! Big thanks to you too. First open source contribution, super happy! haha
